### PR TITLE
Fix #1443 - trim teaEncode trailing zeroes

### LINKED ISSUE #1443: Tea encoding bug
**Describe the bug**

My version: Multi Theft Auto v1.5.7-release-20520
The problem occurs when using the encodeString, decodeString, teaDecode and teaEncode functions. The characters "\0" are added to the decrypted string. Exists only on the client side, and when the number of characters is not a multiple of 4.

**To reproduce**
Use the `iprint(decodeString("tea",encodeString("tea",1,{key="123"}),{key="123"}))`

**Expected behaviour**
Return the correct value

**Screenshots**

none

**Version**

Client Windows
Multi Theft Auto v1.5.7-release-20520
Copyright (C) 2003 - 2020 Multi Theft Auto

**Additional context**

Example:
`iprint(decodeString("tea",encodeString("tea",1,{key="123"}),{key="123"}))`
returns "1\0\0\0" instead of "1"

More examples:
String before encoding - String after decoding

```
"1" - "1\0\0\0"
"12" - "12\0\0"
"123" - 123\0"
"1234" - "1234"
"1234567" - "1234567\0"
"12345678" - "12345678"
```

### PR DESCRIPTION
**Pull request by Pirulax**
So basically, with the introduction of sbx's new parser, we started returning strings with `lua_pushlstring`, which takes a length as an argument, thus pushes the whole string along with padding. The old method used `lua_pushstring` which used `strlen`, thus god rid of the padding.
This PR fixes #1443 .

Edit: test resource: [utf1443.zip](https://github.com/multitheftauto/mtasa-blue/files/4813800/utf1443.zip)

### CODE REVIEW

File: Shared/sdk/SharedUtil.Hash.hpp [hunk 1]
```diff
@@ -818,5 +818,22 @@ namespace SharedUtil
...
+        if (trimPaddingAtEnd)
+        {
+            // Check last block for 0 padding, and delete
+
+            auto iter = out->end() - 1;
+            if (*iter != 0)
+                return;
+
+            const auto blockEnd = out->end() - 4;
+            while (iter > blockEnd && *--iter == 0);
```
Comment (qaisjp) [inline Shared/sdk/SharedUtil.Hash.hpp]: Please use more lines

File: Shared/sdk/SharedUtil.Hash.hpp [hunk 2]
```diff
@@ -772,7 +772,7 @@ namespace SharedUtil
         delete[] strbuf;
     }
 
-    void TeaDecode(const SString& str, const SString& key, SString* out)
+    void TeaDecode(const SString& str, const SString& key, SString* out, const bool trimPaddingAtEnd)
```
Comment (qaisjp) [inline Shared/sdk/SharedUtil.Hash.hpp]: Is this necessary? The only place where this is false is in `Shared/sdk/SharedUtil.Tests.hpp`, so it seems like this is unnecessary.

**Review by qaisjp — DISMISSED** [review]

File: Shared/sdk/SharedUtil.Hash.hpp (hunk 1 above)
Comment (Pirulax) [reply Shared/sdk/SharedUtil.Hash.hpp]: I feel like this is perfectly readable. Not that anyone reads that code anyways.

File: Shared/sdk/SharedUtil.Hash.hpp (hunk 2 above)
Comment (Pirulax) [reply Shared/sdk/SharedUtil.Hash.hpp]: No, it's not.

File: Shared/sdk/SharedUtil.Hash.hpp [hunk 3]
```diff
@@ -818,5 +818,22 @@ namespace SharedUtil
 
         out->assign((char*)buffer, numPasses * 4);
         delete[] buffer;
+
+        if (trimPaddingAtEnd)
+        {
+            // Check last block for 0 padding, and delete
+
+            auto iter = out->end() - 1;
```
Comment (sbx320) [inline Shared/sdk/SharedUtil.Hash.hpp]: What if `out.begin() == out.end()` (aka. the input is empty)?

File: Shared/sdk/SharedUtil.Hash.hpp [hunk 4]
```diff
@@ -818,5 +818,22 @@ namespace SharedUtil
...
+            if (*iter != 0)
+                return;
+
+            const auto blockEnd = out->end() - 4;
```
Comment (sbx320) [inline Shared/sdk/SharedUtil.Hash.hpp]: This doesn't seem correct. If I encode `abc\0\0\0`, then the decode result (pre-trim) would be `abc\0\0\0\0\0`. After trimming it ends up being `abc\0`, which is incorrect. We should trim the exact amount of excess padding that was added previously..

**Review by sbx320 — CHANGES_REQUESTED** [review]

File: Shared/sdk/SharedUtil.Hash.hpp (hunk 4 above)
Comment (ccw808) [reply Shared/sdk/SharedUtil.Hash.hpp]: Previous version did the same thing. I think that should be looked at in a new pull request

File: Shared/sdk/SharedUtil.Hash.hpp (hunk 3 above)
Comment (Pirulax) [reply Shared/sdk/SharedUtil.Hash.hpp]: The block size is 4. If its smaller than that the function returns before ever reaching this point.

See code:
```cpp
        // Count the number of passes that we need
        int numBlocks = str.length() / 4;
        int numPasses = numBlocks - 1;

        if (numPasses <= 0)
            return;
```

File: Shared/sdk/SharedUtil.Hash.hpp (hunk 4 above)
Comment (Pirulax) [reply Shared/sdk/SharedUtil.Hash.hpp:829]: I feel like doing `strlen` on a 10-15 mb string isn't particularly efficient.

Comment (Pirulax) [reply Shared/sdk/SharedUtil.Hash.hpp]: So it this the correct behavior?

Comment (Pirulax) [reply Shared/sdk/SharedUtil.Hash.hpp]: @sbx320  I don't see a way to achieve what you want. We can't possible figure out if it was padding, or the user has added the `\0\0\0`'s. Please correct me if there's a way to do so.

Comment (Pirulax) [reply Shared/sdk/SharedUtil.Hash.hpp]: I can trim all `0`'s, not just the ones in the last block. That would leave us with a string that has no padding. But if the user added `0`'s we can't figure out on the other end if that was padding, or it was originally in the string

**Comment by Pirulax** [comment]
I can't connect to my 1.5.6 server, for some stupid reason(a resource sets min client version, and I cant figure out which one)

Please, could someone run a few tests, and post the results here on 1.5.6 server and client?

Test values:
- 123
- 1234
- \0\0Hello
- 123\0
- hithere\0\0\0\0
- \0hi\0
- \0\0\0\0\0hi\0\0
- 123\0\0\0\0

They key can be anything. Let it be.. "c++20"

**Comment by ccw808** [comment]
@Pirulax mtaserver.conf:
```
    <minclientversion></minclientversion>
    <minclientversion_auto_update>0</minclientversion_auto_update>
```
Edit: Oh a resource sets it - Nevermind

File: Shared/sdk/SharedUtil.Hash.hpp (hunk 1 above)
Comment (qaisjp) [reply Shared/sdk/SharedUtil.Hash.hpp]: This requires less thinking
```
while (iter > blockEnd || *iter == 0)
    iter--;
```

**Comment by Pirulax** [comment]
btw, the problem is that, now I've fixed `en/decodeString` as well, which has been returning wrong values since it was introduced. Should we keep backwards compatibility? Or just hope nobody tries comparing decoded strings?

**Comment by Pirulax** [comment]
Btw, since the old one used `strnlen` it strips all padding from the end, so I'll modify this one to be like that.

**Comment by Pirulax** [comment]
> @Pirulax mtaserver.conf:
> 
> ```
>     <minclientversion></minclientversion>
>     <minclientversion_auto_update>0</minclientversion_auto_update>
> ```
> 
> Edit: Oh a resource sets it - Nevermind

Actually, I'm not sure if it was a resource. I think this'll help, thanks!

**Comment by Pirulax** [comment]
New resource to test: [utf1443.zip](https://github.com/multitheftauto/mtasa-blue/files/4819898/utf1443.zip)

File: Shared/sdk/SharedUtil.Hash.hpp [hunk 5]
```diff
@@ -818,5 +818,26 @@ namespace SharedUtil
 
         out->assign((char*)buffer, numPasses * 4);
         delete[] buffer;
+
+
```
Comment (qaisjp) [inline Shared/sdk/SharedUtil.Hash.hpp:822]:
```suggestion
```

File: Shared/sdk/SharedUtil.Hash.hpp [hunk 6]
```diff
@@ -696,7 +696,7 @@ namespace SharedUtil
             return GenerateHashHexString(hashFunction, NULL, 0);
     }
 
-    void encodeXtea(unsigned int* v, unsigned int* w, unsigned int* k)
+    inline void encodeXtea(unsigned int* v, unsigned int* w, unsigned int* k)
```
Comment (qaisjp) [inline Shared/sdk/SharedUtil.Hash.hpp:699]: This inlining is not relevant to the fix.

**Comment by Pirulax** [comment]
We need to fix the padding in teaEncode, because with binary files 0's at the file end is totally acceptable.

File: Shared/sdk/SharedUtil.Hash.hpp (hunk 6 above)
Comment (qaisjp) [reply Shared/sdk/SharedUtil.Hash.hpp:699]:
```suggestion
   void encodeXtea(unsigned int* v, unsigned int* w, unsigned int* k)
```

File: Shared/sdk/SharedUtil.Hash.hpp [hunk 7]
```diff
@@ -696,7 +696,7 @@ namespace SharedUtil
             return GenerateHashHexString(hashFunction, NULL, 0);
     }
 
-    void encodeXtea(unsigned int* v, unsigned int* w, unsigned int* k)
+   void encodeXtea(unsigned int* v, unsigned int* w, unsigned int* k)
```
Comment (qaisjp) [inline Shared/sdk/SharedUtil.Hash.hpp]:
```suggestion
    void encodeXtea(unsigned int* v, unsigned int* w, unsigned int* k)
```

File: Shared/sdk/SharedUtil.Hash.hpp [hunk 8]
```diff
@@ -710,7 +710,7 @@ namespace SharedUtil
         w[1] = v1;
     }
 
-    void decodeXtea(unsigned int* v, unsigned int* w, unsigned int* k)
+    inline void decodeXtea(unsigned int* v, unsigned int* w, unsigned int* k)
```
Comment (qaisjp) [inline Shared/sdk/SharedUtil.Hash.hpp:713]:
```suggestion
    void decodeXtea(unsigned int* v, unsigned int* w, unsigned int* k)
```

**Comment by Pirulax** [comment]
Perhaps, maybe add a noTrim argument?

**Comment by qaisjp** [comment]
Lets try and fix the regression in this PR and we can discuss other problems with teaEncode in a separate issue.

**Comment by ccw808** [comment]
> 
> 
> Lets try and fix the regression in this PR and we can discuss other problems with teaEncode in a separate issue.

Agreed. Let's get this fixed yesterday

**Pull request merged**